src/interface/sparse_pack.hpp

```diff
@@ -0,0 +1,282 @@
+struct GetTypeIdx;
+
+template <typename T, typename... Ts>
+struct GetTypeIdx<T, T, Ts...> : std::integral_constant<std::size_t, 0> {
+  using type = void;
+};
+
+template <typename T, typename U, typename... Ts>
+struct GetTypeIdx<T, U, Ts...>
+    : std::integral_constant<std::size_t, 1 + GetTypeIdx<T, Ts...>::value> {
+  using type = void;
+};
+
+template <class T, class... Ts>
+struct IncludesType;
+
+template <typename T>
+struct IncludesType<T, T> : std::true_type {};
+
+template <typename T, typename... Ts>
+struct IncludesType<T, T, Ts...> : std::true_type {};
+
+template <typename T, typename U>
+struct IncludesType<T, U> : std::false_type {};
+
+template <typename T, typename U, typename... Ts>
+struct IncludesType<T, U, Ts...> : IncludesType<T, Ts...> {};
+
+} // namespace impl
+
+using namespace impl;
+
+namespace variables {
+// Struct that all variables types should inherit from
+template <bool REGEX, int... NCOMP>
+struct base_t {
+  KOKKOS_INLINE_FUNCTION
+  base_t() : idx(0) {}
+
+  KOKKOS_INLINE_FUNCTION
+  explicit base_t(int idx1) : idx(idx1) {}
+
+  virtual ~base_t() = default;
+
+  // All of these are just static methods so that there is no
+  // extra storage in the struct
+  static std::string name() {
+    PARTHENON_FAIL("Need to implement your own name method.");
+    return "error";
+  }
+  KOKKOS_INLINE_FUNCTION
+  static bool regex() { return REGEX; }
+  KOKKOS_INLINE_FUNCTION
+  static int ndim() { return sizeof...(NCOMP); }
+  KOKKOS_INLINE_FUNCTION
+  static int size() { return multiply<NCOMP...>::value; }
+
+  const int idx;
+};
+
+struct any : public base_t<true> {
+  template <class... Ts>
+  KOKKOS_INLINE_FUNCTION any(Ts &&...args)
+      : parthenon::variables::base_t<true>(std::forward<Ts>(args)...) {}
+  static std::string name() { return ".*"; }
+};
+} // namespace variables
+
+template <class... Ts>
+class SparsePack : public SparsePackBase {
+ public:
+  SparsePack() = default;
+
+  explicit SparsePack(const SparsePackBase &spb) : SparsePackBase(spb) {}
+
+  template <class T>
+  static SparsePack Make(T *pmd, const std::vector<MetadataFlag> &flags = {},
+                         bool fluxes = false, bool coarse = false) {
+    auto &cache = pmd->GetSparsePackCache();
+    return SparsePack(cache.Get(
+        pmd, PackDescriptor(std::vector<std::string>{Ts::name()...},
+                            std::vector<bool>{Ts::regex()...}, flags, fluxes, coarse)));
+  }
+
+  template <class T, class VAR_VEC>
+  static std::tuple<SparsePack, SparsePackIdxMap>
+  Make(T *pmd, const VAR_VEC &vars, const std::vector<MetadataFlag> &flags = {},
+       bool fluxes = false, bool coarse = false) {
+    static_assert(sizeof...(Ts) == 0);
+    auto out = SparsePackBase::Make(pmd, vars, flags, fluxes, coarse);
+    return {SparsePack(std::get<0>(out)), std::get<1>(out)};
+  }
+
+  template <class T>
+  static SparsePack MakeWithFluxes(T *pmd, const std::vector<MetadataFlag> &flags = {}) {
```

Comment:
Personally, I think it makes for a cleaner interface since I think it can be confusing when a function takes multiple booleans  (since reading `func(true, false, true)` is often unclear). Additionally, I think we want to prevent people from makeing packs with fluxes and coarse variables unless they really want to.